examples/features/captures.effekt.md

```diff
@@ -0,0 +1,135 @@
+# Captures
+
+## Notes (TO BE IGNORED)
+
+Boxing is almost never needed. The cases where you need it are complex yet interesting. While it appears to be complicated, it guarantees safety and helps writing correct programs.
+
+Aspects
+
+- disallowing effects (only `{ io, global }` since these can be run without a handler or pure `{}`)
+  - FFI on click (event-handler)
+  ```
+  do promise(box { println("hello") })
+  ```
+- storing the resumption/continuation (scheduler, reified generator)
+- classes-pattern
+  ```
+  interface Counter {
+    def tick(): Unit
+    def current(): Int
+  }
+
+  def newCounter(init: Int): Counter at {global} = {
+    var state in global = init // or pass a region
+    box new Counter {
+      def tick() = state = state + 1
+      def current() = state
+    }
+  }
+  /*
+  class GlobalCounter(init: Int): Counter {
+    var state in global = init // or pass a region
+    def tick() = state = state + 1
+    def current() = state
+  }
+  */
+
+  def myCounter = unbox newCounter(0)
+  ```
+  alternative we can write it in "CPS" to avoid boxing
+  ```
+  def counter[R](init: Int) { prog: {Counter} => R }: R = {
+    var state = init // NO region needed here
+    def c = new Counter {
+      def tick() = state = state + 1
+      def current() = state
+    }
+    prog {c}
+  }
+
+  // ...
+  with def myCounter = counter(0);
+  ...
+  ```
+  - on term level unbox, box
+  - `at` at type level
+  - capability, resource, region, lifetime, coeffect
+
+## Captures
+
+Functions are second-class in Effekt, but for a good reason. Functions closing over capabilities must not be returned, otherwise it can not be guaranteed that these capabilities are still in scope.
+Thus, we explicitly have to keep track of these captures and ensure they are in scope upon envoking.
+
+Consider the following example where `divide` is being passed a explicit capability:
+
+```
+def divide(n: Int, m: Int) {exc: Exception}: Int =
+  if (m == 0) { exc.throw("uhoh") }
+  else { n / m }
+```
+
+```effekt:repl
+try { 
+  val res = divide(10, 0) {exc}
+  println(res)
+} with exc: Exception {  
+  def throw(msg) = println("ERROR: " ++ show(msg))
+}
+```
+
+What happens if we return the capability?
+
+```
+def example() = try {
+  exc
+} with exc: Exception { 
+  def throw(msg) = panic("ohno")
+}
+```
+
+By directly returning `exc` from the `try` expression, we break the lexical effect handling reasoning since `exc` is only defined within the `try` body.
+For the same reason, returning a closure is not an option either.
+
+For preventing capabilities escaping their scope, Effekt uses captures. In case of the previous example, the return type of `example` is `Exception at {exc}`. This is called
+a boxed computation. A boxed computation may only be used after unboxing it, which is only permitted when all its captures are in scope.
```

Comment:
```suggestion
To prevent capabilities escaping their scope, Effekt tracks their capture. 
In case of the previous example, the return type of `example` would be `Exception at {exc}`. 
That is, it expresses that the capability of type `Exception` can only be used when `exc` is in scope. 
The Effekt type checker prevents this example from type checking since `exc` is not even a valid name outside of its binding effect handler.
```